src/main/java/com/microsoft/sqlserver/jdbc/SQLQueryMetadataCache.java

```diff
@@ -0,0 +1,281 @@
+
+            /*
+             * If ever the map doesn't contain a parameter, the cache entry cannot be used. If there is data found, it
+             * should never have the initialized algorithm as that would contain the key. Clear all metadata that has
+             * already been assigned in either case.
+             */
+            if (!found || (foundData != null && foundData.isAlgorithmInitialized())) {
+                for (Parameter param : params) {
+                    param.cryptoMeta = null;
+                }
+                return false;
+            }
+
+            params[i].cryptoMeta = foundData;
+        }
+
+        // Assign the key using a metadata copy. We shouldn't load from the cached version for security reasons.
+        for (int i = 0; i < params.length; ++i) {
+            try {
+                CryptoMetadata cryptoCopy = null;
+                CryptoMetadata metaData = params[i].getCryptoMetadata();
+                if (metaData != null) {
+                    cryptoCopy = new CryptoMetadata(metaData.getCekTableEntry(), metaData.getOrdinal(),
+                            metaData.getEncryptionAlgorithmId(), metaData.getEncryptionAlgorithmName(),
+                            metaData.getEncryptionType().getValue(), metaData.getNormalizationRuleVersion());
+                }
+
+                params[i].cryptoMeta = cryptoCopy;
+
+                if (cryptoCopy != null) {
+                    try {
+                        SQLServerSecurityUtility.decryptSymmetricKey(cryptoCopy, connection, stmt);
+                    } catch (SQLServerException e) {
+
+                        removeCacheEntry(stmt, session, connection);
+
+                        for (Parameter paramToCleanup : params) {
+                            paramToCleanup.cryptoMeta = null;
+                        }
+
+                        return false;
+                    }
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+
+        }
+        
+        Map<Integer, CekTableEntry> enclaveKeys = session.getCryptoCache().getEnclaveEntry(encryptionValues.getValue());
+        return (enclaveKeys == null);
+    }
+
+    
+    /**
+    * 
+    * @param params
+    *        List of parameters used
+    * @param parameterNames
+    *        Names of parameters used
+    * @param session
+    *        Enclave session containing the cryptocache
+    * @param connection
+    *        SQLServerConnection
+    * @param stmt
+    *        SQLServer statement used to retrieve keys to find correct cache
+    * @param cekList
+    *        The list of CEKs (from the first RS) that is also added to the cache as well as parameter metadata
+    * @return true, if the query metadata has been added correctly
+*/
+    public static boolean addQueryMetadata(Parameter[] params, ArrayList<String> parameterNames, EnclaveSession session,

```

Comment:
these shouldn't be public

Reply:
All methods set to default. They were initially public as I thought they would be needed in tests, but that didn't turn out to be the case.